unit_tests/log_transfer/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,872 @@
+
+  return pc;
+}
+
+CLIENTIDS
+test_env_chkpt::generate_client (int index)
+{
+  clientids clnt;
+  char number[25];
+
+  clnt.client_type = static_cast<db_client_type> (1);
+  sprintf (number, "client_%d", index);
+  clnt.client_info = (number);
+  sprintf (number, "db_user_%d", index);
+  clnt.db_user = (number);
+  sprintf (number, "database_%d", index);
+  clnt.program_name = (number);
+  sprintf (number, "login_%d", index);
+  clnt.login_name = (number);
+  sprintf (number, "host_%d", index);
+  clnt.host_name = (number);
+  clnt.process_id = rand () % MAX_RAND;
+
+  return clnt;
+}
+
+LOG_TDES *
+test_env_chkpt::generate_tdes (int index)
+{
+  LOG_TDES *tdes = new log_tdes();
+  tdes->trid = index + 1;
+  tdes->tran_index = index + 1;
+
+  tdes->tail_lsa = generate_log_lsa();
+  tdes->isloose_end = std::rand() % 2;
+  tdes->head_lsa = generate_log_lsa();
+  if (index < TRAN_UNACTIVE_UNKNOWN)
+    {
+      tdes->state    = static_cast<TRAN_STATE> (index);
+    }
+  else
+    {
+      tdes->state    = static_cast<TRAN_STATE> (std::rand() % TRAN_UNACTIVE_UNKNOWN);
+    }
+
+
+  tdes->undo_nxlsa = generate_log_lsa();
+  tdes->posp_nxlsa = generate_log_lsa();
+  tdes->savept_lsa = generate_log_lsa();
+  tdes->tail_topresult_lsa = generate_log_lsa();
+  tdes->rcv.tran_start_postpone_lsa = generate_log_lsa();
+  tdes->wait_msecs = rand() % MAX_RAND;
+  tdes->client_id  = rand() % MAX_RAND;
+  tdes->gtrid      = rand() % MAX_RAND;
+
+  tdes->num_transient_classnames = rand() % MAX_RAND;
+  tdes->num_repl_records         = rand() % MAX_RAND;
+  tdes->cur_repl_record          = rand() % MAX_RAND;
+  tdes->append_repl_recidx       = rand() % MAX_RAND;
+  tdes->fl_mark_repl_recidx      = rand() % MAX_RAND;
+  tdes->repl_insert_lsa          = generate_log_lsa();
+  tdes->repl_update_lsa          = generate_log_lsa();
+
+  tdes->client  = generate_client (index);
+  tdes->gtrinfo = generate_2pc_gtrinfo();
+  tdes->coord   = generate_2pc_coordinator();
+
+  return tdes;
+}
+
+
+void
+test_env_chkpt::generate_tran_table()
+{
+  log_Gl.trantable.num_total_indices = 120;
+
+  int size = log_Gl.trantable.num_total_indices * sizeof (*log_Gl.trantable.all_tdes);
+  log_Gl.trantable.all_tdes = (LOG_TDES **) malloc (size);
+
+  for (int i = 1; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      log_Gl.trantable.all_tdes[i] = generate_tdes (i);
+    }
+
+  log_Gl.trantable.all_tdes[101]->tail_lsa = NULL_LSA;
+  log_Gl.trantable.all_tdes[101]->trid = NULL_TRANID;
+
+  log_Gl.trantable.all_tdes[102]->tail_lsa = NULL_LSA;
+  log_Gl.trantable.all_tdes[102]->trid = NULL_TRANID;
+
+  log_Gl.trantable.all_tdes[103]->tail_lsa = NULL_LSA;
+  log_Gl.trantable.all_tdes[103]->trid = NULL_TRANID;
+
+  log_Gl.trantable.all_tdes[104]->trid = -104;
+  log_Gl.trantable.all_tdes[104]->rcv.atomic_sysop_start_lsa = generate_log_lsa ();
+  log_Gl.trantable.all_tdes[105]->trid = -105;
+  log_Gl.trantable.all_tdes[105]->rcv.sysop_start_postpone_lsa = generate_log_lsa ();
```

Comment:
Let's try to make a more precise generation :). Also, both trantable and systb_System_tdes need to be generated to test checkpoint info completely.

```suggestion
  log_Gl.trantable.num_total_indices = 18;
  
  log_tdes *tdes;
  int tran_index = 1;
  
  // Generate with NULL_TRANID
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->trid = NULL_TRANID;
  ++tran_index;
  
  // Generate with NULL tail
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->tail_lsa = NULL_LSA;
  ++tran_index;
  
  // Generate with (almost) every state
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_ACTIVE;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_COMMITTED;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_WILL_COMMIT;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_COMMITTED_WITH_POSTPONE;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_ABORTED;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_UNILATERALLY_ABORTED;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_2PC_PREPARE;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_2PC_COLLECTING_PARTICIPANT_VOTES;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_2PC_ABORT_DECISION;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_2PC_COMMIT_DECISION;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_COMMITTED_INFORMING_PARTICIPANTS;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->state = TRAN_UNACTIVE_ABORTED_INFORMING_PARTICIPANTS;
  ++tran_index;

  // Generate topops
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->topops.last = 0;
  tdes->rcv.sysop_start_postpone_lsa = NULL_LSA;
  tdes->rcv.atomic_sysop_start_lsa = NULL_LSA;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->topops.last = 1;
  tdes->rcv.sysop_start_postpone_lsa = {1, 1};
  tdes->rcv.atomic_sysop_start_lsa = NULL_LSA;
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->topops.last = 2;
  tdes->rcv.sysop_start_postpone_lsa = NULL_LSA;
  tdes->rcv.atomic_sysop_start_lsa = {1, 1};
  ++tran_index;
  log_Gl.trantable.alltdes[tran_index] = tdes = generate_tdes (tran_index);
  tdes->topops.last = 3;
  tdes->rcv.sysop_start_postpone_lsa = {1, 1};
  tdes->rcv.atomic_sysop_start_lsa = {1, 1};
  ++tran_index;
  
  assert (tran_index == log_Gl.trantable.num_total_indices);
  
  // Generate system tdes with and without topops
  tran_index = -2;
  tdes = generate_tdes (tran_index);
  tdes->topops.last = -1;
  systb_System_tdes[tran_index] = tdes;

  tran_index = -3;
  tdes = generate_tdes (tran_index);
  tdes->topops.last = 0;
  tdes->rcv.sysop_start_postpone_lsa = {1, 1};
  tdes->rcv.atomic_sysop_start_lsa = {1, 1};
  systb_System_tdes[tran_index] = tdes;
```